File: src/simulation/culture/group/centers/CultureAspectCenter.java
package simulation.culture.group.centers;


import kotlin.Pair;
import simulation.culture.group.cultureaspect.*;
import simulation.culture.group.reason.ConstructReasonsKt;
import simulation.culture.group.reason.Reason;
import simulation.culture.thinking.meaning.ConstructMemeKt;
import simulation.space.resource.Resource;

import java.util.*;
import java.util.function.Predicate;
import java.util.stream.Collectors;

import static shmp.random.RandomCollectionsKt.randomElement;
import static shmp.random.RandomProbabilitiesKt.testProbability;
import static simulation.Controller.session;

public class CultureAspectCenter {
    private Group group;
    private MutableCultureAspectPool aspectPool;
    private Set<Resource> aestheticallyPleasingResources = new HashSet<>();

    private Set<Reason> reasonsWithSystems = new HashSet<>();

    CultureAspectCenter(Group group, Set<CultureAspect> cultureAspects) {
        this.group = group;
        this.aspectPool = new MutableCultureAspectPool(cultureAspects);
    }

    public MutableCultureAspectPool getAspectPool() {
        return aspectPool;
    }

    public void addCultureAspect(CultureAspect cultureAspect) {
        if (cultureAspect != null) {
            aspectPool.add(cultureAspect);
        }
        if (cultureAspect instanceof AestheticallyPleasingObject) {
            aestheticallyPleasingResources.add(
                    ((AestheticallyPleasingObject) cultureAspect).getResource()
            );
        }
    }

    void useCultureAspects() {
        aspectPool.getAll().forEach(a -> a.use(group));
    }

    void addRandomCultureAspect(Group group) {
        if (!testProbability(session.cultureAspectBaseProbability, session.random)) {
            return;
        }
        CultureAspect cultureAspect = null;
        switch (session.random.nextInt(5)) {
            case 0: {
                cultureAspect = ConstructCultureAspectKt.createDepictObject(
                        group.getCultureCenter().getAspectCenter().getAspectPool().getMeaningAspects(),
                        ConstructMemeKt.constructAndAddSimpleMeme(
                                group.getCultureCenter().getMemePool(),
                                session.random,
                                0.5
                        ),
                        group,
                        session.random
                );
                break;
            }
            case 2: {
                cultureAspect = ConstructCultureAspectKt.createAestheticallyPleasingObject(
                        group.getCultureCenter().getAspectCenter().getAspectPool().getProducedResources()
                                .stream()
                                .map(Pair::getFirst)
                                .filter(r -> !aestheticallyPleasingResources.contains(r))
                                .max(Comparator.comparingInt(r -> r.getGenome().getBaseDesirability()))
                                .orElse(null),
                        group,
                        session.random
                );
                break;
            }
            case 3: {//TODO recursively go in dependencies;
                cultureAspect = ConstructCultureAspectKt.constructRitual(
                        ConstructReasonsKt.constructBetterAspectUseReason(
                                group,
                                group.getCultureCenter().getAspectCenter().getAspectPool().getConverseWrappers(),
                                reasonsWithSystems,
                                session.random,
                                5//TODO meh, java
                        ),
                        group,
                        session.random
                );
                break;
            }
            case 4: {
                cultureAspect = ConstructCultureAspectKt.createTale(
                        group,
                        session.templateBase,
                        session.random
                );
                break;
            }
        }
        addCultureAspect(cultureAspect);
    }

    void mutateCultureAspects() {
        if (!testProbability(session.groupCultureAspectCollapse, session.random)) {
            return;
        }
        switch (session.random.nextInt(3)) {
            case 0: {
                joinSimilarRituals();
                break;
            }
            case 1: {
                joinSimilarTalesBy("!actor");
//                joinSimilarTalesBy("@verb");//TODO debug off
                break;
            }
            case 2: {
                addCultureAspect(
                        ChangeCultureAspectsKt.takeOutDeity(aspectPool, session.random)
                );
                break;
            }
        }
    }

    void joinSimilarRituals() {
        RitualSystem system = ChangeCultureAspectsKt.takeOutSimilarRituals(aspectPool, group, 3);
        if (system != null) {
            addCultureAspect(system);
            reasonsWithSystems.add(system.getReason());
        }
    }

    void joinSimilarTalesBy(String infoTag) {
        TaleSystem system = ChangeCultureAspectsKt.takeOutSimilarTalesByTag(infoTag, aspectPool, 3);
        if (system != null) {
            addCultureAspect(system);
        }
    }

    public List<Pair<CultureAspect, Group>> getNeighbourCultureAspects() {
        return group.getRelationCenter().getRelatedGroups().stream()
                .flatMap(g ->
                        g.getCultureCenter().getCultureAspectCenter().getAspectPool().getAll().stream()
                                .map(a -> new Pair<>(a, g))
                ).collect(Collectors.toList());
    }

    public List<Pair<CultureAspect, Group>> getNeighbourCultureAspects(Predicate<CultureAspect> predicate) {
        return getNeighbourCultureAspects().stream()
                .filter(p -> predicate.test(p.getFirst()))
                .collect(Collectors.toList());
    }

    void adoptCultureAspects(Group group) {
        if (!session.isTime(session.groupTurnsBetweenAdopts)) {
            return;
        }
        List<Pair<CultureAspect, Group>> cultureAspects = getNeighbourCultureAspects(a -> !aspectPool.contains(a));
        if (!cultureAspects.isEmpty()) {
            try {
                CultureAspect aspect = randomElement(
                        cultureAspects,
                        a -> group.getRelationCenter().getNormalizedRelation(a.getSecond()),
                        session.random
                ).getFirst().copy(group);
                addCultureAspect(aspect);
            } catch (NoSuchElementException e) {}//TODO mb some more smart check?
        }
    }
}

File: src/simulation/culture/group/centers/CultureAspectCenter.kt
package simulation.culture.group.centers

import shmp.random.SampleSpaceObject
import shmp.random.randomElement
import shmp.random.testProbability
import simulation.Controller.*
import simulation.culture.group.cultureaspect.*
import simulation.culture.group.cultureaspect.worship.Worship
import simulation.culture.group.reason.Reason
import simulation.culture.group.reason.constructBetterAspectUseReason
import simulation.culture.thinking.meaning.constructAndAddSimpleMeme
import simulation.space.resource.Resource
import java.util.*
import kotlin.math.pow

class CultureAspectCenter(private val group: Group) {
    val aspectPool = MutableCultureAspectPool(mutableSetOf())
    private val aestheticallyPleasingResources: MutableSet<Resource> = HashSet()
    private val reasonsWithSystems: MutableSet<Reason> = HashSet()

    fun addCultureAspect(cultureAspect: CultureAspect?) {
        cultureAspect ?: return
        aspectPool.add(cultureAspect)
        if (cultureAspect is CherishedResource)
            aestheticallyPleasingResources.add(cultureAspect.resource)
    }

    fun useCultureAspects() = aspectPool.all.forEach { it.use(group) }

    fun addRandomCultureAspect(group: Group) {
        if (!testProbability(session.cultureAspectBaseProbability, session.random))
            return
        var cultureAspect: CultureAspect? = null
        when (randomElement(AspectRandom.values(), session.random)) {
            AspectRandom.Depict -> cultureAspect = createDepictObject(
                    group.cultureCenter.aspectCenter.aspectPool.getMeaningAspects(),
                    constructAndAddSimpleMeme(
                            group.cultureCenter.memePool,
                            session.random
                    ),
                    group,
                    session.random
            )
            AspectRandom.AestheticallyPleasing -> cultureAspect = createAestheticallyPleasingObject(
                    group.cultureCenter.aspectCenter.aspectPool.producedResources
                            .filter { it.genome.isDesirable }
                            .filter { !aestheticallyPleasingResources.contains(it) }
                            .maxBy { it.genome.baseDesirability },
                    group,
                    session.random
            )
            AspectRandom.Ritual -> cultureAspect = constructRitual(//TODO recursively go in dependencies;
                    constructBetterAspectUseReason(
                            group,
                            group.cultureCenter.aspectCenter.aspectPool.converseWrappers,
                            reasonsWithSystems,
                            session.random
                    ),
                    group,
                    session.random
            )
            AspectRandom.Tale -> cultureAspect = createTale(
                    group,
                    session.templateBase,
                    session.random
            )
        }
        addCultureAspect(cultureAspect)
    }

    fun mutateCultureAspects(group: Group) {
        if (!testProbability(session.groupCultureAspectCollapse, session.random))
            return
        when (randomElement(ChangeRandom.values(), session.random)) {
            ChangeRandom.RitualSystem -> joinSimilarRituals()
            ChangeRandom.TaleSystem -> joinSimilarTalesBy("!actor")
            ChangeRandom.Worship -> addCultureAspect(takeOutWorship(aspectPool, session.random))
            ChangeRandom.God -> makeGod(group)
        }
    }

    private fun joinSimilarRituals() {
        val system = takeOutSimilarRituals(aspectPool) ?: return
        addCultureAspect(system)
        reasonsWithSystems.add(system.reason)
    }

    private fun makeGod(group: Group) {
        val cult = takeOutGod(aspectPool, group, session.random) ?: return
        addCultureAspect(cult)
    }

    private fun joinSimilarTalesBy(infoTag: String) {
        val system = takeOutSimilarTalesByTag(infoTag, aspectPool)
        system?.let { addCultureAspect(it) }
    }

    private val neighbourCultureAspects: List<Pair<CultureAspect, Group>>
        get() = group.relationCenter.relatedGroups.flatMap { n ->
            n.cultureCenter.cultureAspectCenter.aspectPool.all.map { a -> Pair(a, n) }
        }

    private fun getNeighbourCultureAspects(predicate: (CultureAspect) -> Boolean): List<Pair<CultureAspect, Group>> =
            neighbourCultureAspects.filter { (f) -> predicate(f) }

    fun adoptCultureAspects(group: Group) {
        if (!session.isTime(session.groupTurnsBetweenAdopts)) return
        val cultureAspects = getNeighbourCultureAspects { !aspectPool.contains(it) }
        //TODO mb some more smart check?
        if (cultureAspects.isNotEmpty()) try {
            val aspect = randomElement(
                    cultureAspects,
                    { (_, g) -> group.relationCenter.getNormalizedRelation(g) },
                    session.random
            ).first.adopt(group)
                    ?: return
            if (shouldAdopt(aspect))
                addCultureAspect(aspect)
        } catch (e: NoSuchElementException) {
        }
    }

    private fun shouldAdopt(aspect: CultureAspect): Boolean {
        if (aspect is Worship) {
            val similarGodsAmount = aspectPool.worships.count { w ->
                w.worshipObject.memes.any { it in aspect.worshipObject.memes }
            }
            val probability = 1.0 / (similarGodsAmount + 1)
            return testProbability(probability.pow(2), session.random)
        }
        return true
    }

    fun die(group: Group) = aspectPool.all.forEach { it.die(group) }
}

private enum class AspectRandom(override val probability: Double) : SampleSpaceObject {
    Depict(1.0),
    AestheticallyPleasing(1.0),
    Ritual(1.0),
    Tale(3.0)
}

private enum class ChangeRandom(override val probability: Double) : SampleSpaceObject {
    RitualSystem(3.0),
    TaleSystem(3.0),
    Worship(2.0),
    God(1.0),
}